### PR TITLE
Watcher: Only trigger a watch if new or schedule/changed

### PR DESCRIPTION
**Pull request by spinscale**
The trigger engine did always create a new schedule data structure, when
the watcher indexing listener called an add. However the indexing
listener also called add, when the watch status was updated. This means,
that upon a watch status update the watch got retriggered, potentially
waiting a defined interval from the watch status update onwards, instead
of waiting from the last run.

This commit only updates the schedule in the trigger engine, if it
actually has changed, otherwise the existing schedule will not be
touched. This has two results

1. If a watch is updated by an execution, the existing interval will not
be touched (meaning the scheduled time will not move forward).
2. If a watch is updated by a user, but the schedule is not changed, it
will not be reset from the update (for example starting to count from 5
minutes again, if the interval was set to 5 minutes).

Furthermore some minor cleanups were applied, making variables final in
the ctor, preventing double creation of variables, preventing shadow naming.

### CODE REVIEW

**Review by hub-cap — APPROVED** [review]

File: x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/trigger/schedule/engine/TickerScheduleEngineTests.java [hunk 1]
```diff
@@ -254,6 +254,29 @@ public void accept(Iterable<TriggerEvent> events) {
...
 
+    public void testAddOnlyWithNewSchedule() {
+        engine.start(Collections.emptySet());
+        final List<TriggerEvent> triggerEvents = new ArrayList<>();
+        engine.register(new Consumer<Iterable<TriggerEvent>>() {
+            @Override
+            public void accept(Iterable<TriggerEvent> events) {
+                events.forEach(triggerEvents::add);
+            }
+        });
```
Comment (jakelandis) [inline x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/trigger/schedule/engine/TickerScheduleEngineTests.java]: minor: Is the register and triggerEvents needed here ?

**Review by jakelandis — APPROVED** [review]
A couple minor/questions but LGTM as-is too.

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/trigger/schedule/engine/TickerScheduleTriggerEngine.java [hunk 2]
```diff
@@ -98,10 +105,10 @@ void checkJobs() {
         for (ActiveSchedule schedule : schedules.values()) {
             long scheduledTime = schedule.check(triggeredTime);
             if (scheduledTime > 0) {
-                logger.debug("triggered job [{}] at [{}] (scheduled time was [{}])", schedule.name,
-                    new DateTime(triggeredTime, UTC), new DateTime(scheduledTime, UTC));
-                events.add(new ScheduleTriggerEvent(schedule.name, new DateTime(triggeredTime, UTC),
-                    new DateTime(scheduledTime, UTC)));
+                DateTime triggeredDateTime = new DateTime(triggeredTime, UTC);
```
[the remark this answers is not in the document]
Comment (spinscale) [reply x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/trigger/schedule/engine/TickerScheduleTriggerEngine.java]: the datetime is used two lines below in the CTOR. We will work on switching fully to java time in the future.

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/trigger/schedule/engine/TickerScheduleTriggerEngine.java [hunk 3]
```diff
@@ -84,7 +84,14 @@ public synchronized void pauseExecution() {
     public void add(Watch watch) {
         assert watch.trigger() instanceof ScheduleTrigger;
         ScheduleTrigger trigger = (ScheduleTrigger) watch.trigger();
-        schedules.put(watch.id(), new ActiveSchedule(watch.id(), trigger.getSchedule(), clock.millis()));
+        ActiveSchedule currentSchedule = schedules.get(watch.id());
+        // only update the schedules data structure if the scheduled trigger really has changed, otherwise the time would be reset again
+        // resulting in later executions, as the time would only count after a watch has been stored, as this code is triggered by the
+        // watcher indexing listener
+        // this also means that updating an existing watch would not retrigger the schedule time, if it remains the same schedule
+        if (currentSchedule == null || currentSchedule.schedule.equals(trigger.getSchedule()) == false) {
```
[the remark this answers is not in the document]
Comment (spinscale) [reply x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/trigger/schedule/engine/TickerScheduleTriggerEngine.java]: indeed. the only way to really fix that would be to implement sth like `Comparable`, but that is not really needed from a functional perspective. I have added a comment for now.

File: x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/trigger/schedule/engine/TickerScheduleEngineTests.java (hunk 1 above)
Comment (spinscale) [reply x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/trigger/schedule/engine/TickerScheduleEngineTests.java]: indeed. removed.

**Pull request merged**